q2_phylogenomics/_filter.py

```diff
@@ -0,0 +1,131 @@
+}
+
+
+def bowtie2_build(sequences: DNAFASTAFormat,
+                  n_threads: str = 1) -> Bowtie2IndexDirFmt:
+    database = Bowtie2IndexDirFmt()
+    build_cmd = ['bowtie2-build', '--threads', str(n_threads),
+                 str(sequences), str(database.path / 'db')]
+    run_command(build_cmd)
+    return database
+
+
+def filter_single(
+        demultiplexed_sequences: SingleLanePerSampleSingleEndFastqDirFmt,
+        database: Bowtie2IndexDirFmt,
+        n_threads: int = _filter_defaults['n_threads'],
+        mode: str = _filter_defaults['mode'],
+        exclude_seqs: bool = _filter_defaults['exclude_seqs']) -> \
+            CasavaOneEightSingleLanePerSampleDirFmt:
+    filtered_seqs = CasavaOneEightSingleLanePerSampleDirFmt()
+    df = demultiplexed_sequences.manifest.view(pd.DataFrame)
+    for _, fwd in df.itertuples():
+        _bowtie2_filter(
+            fwd, None, filtered_seqs, database, n_threads, mode, exclude_seqs)
+    return filtered_seqs
+
+
+def filter_paired(
+        demultiplexed_sequences: SingleLanePerSamplePairedEndFastqDirFmt,
+        database: Bowtie2IndexDirFmt,
+        n_threads: int = _filter_defaults['n_threads'],
+        mode: str = _filter_defaults['mode'],
+        exclude_seqs: bool = _filter_defaults['exclude_seqs']) -> \
+            CasavaOneEightSingleLanePerSampleDirFmt:
+    filtered_seqs = CasavaOneEightSingleLanePerSampleDirFmt()
+    df = demultiplexed_sequences.manifest.view(pd.DataFrame)
+    for _, fwd, rev in df.itertuples():
+        _bowtie2_filter(
+            fwd, rev, filtered_seqs, database, n_threads, mode, exclude_seqs)
+    return filtered_seqs
+
+
+def _bowtie2_filter(f_read, r_read, outdir, database, n_threads, mode,
+                    exclude_seqs):
+    mode = '--' + mode.replace('-global', '')
+    with tempfile.NamedTemporaryFile() as sam_f:
+        samfile_output_path = sam_f.name
+        with tempfile.NamedTemporaryFile() as bam_f:
+            bamfile_output_path = bam_f.name
+
+            # align to reference with bowtie
+            bowtie_cmd = ['bowtie2', '-p', str(n_threads), mode,
+                          '-x', str(database.path / database.get_basename())]
+            if r_read is not None:
+                bowtie_cmd += ['-1', f_read, '-2', r_read]
+            else:
+                bowtie_cmd += ['-U', f_read]
+            bowtie_cmd += ['-S', samfile_output_path]
+            run_command(bowtie_cmd)
+
+            # Filter alignment and convert to BAM with samtools
+            # -f 4 keeps only single alignments that are unmapped
+            # -f 12 keeps only paired alignments with both reads unmapped
+            # -F 256 removes reads that are not primary alignment
+            # -F 260 removes reads that are not primary alignment or unmapped
+            # -F 268 removes reads that are not primary alignment or unmapped
+            # or pair is unmapped.
+            if exclude_seqs:
+                sam_flags = ['-F', '256', '-f', '4']
```

Comment:
would you mind doing the bitwise ops with UPPERCASE_CONSTANTS at the module level? It would make it easier to verify or add things as needed, samtools flags appear to be kind of involved.

Reply:
👍